[rpc] refactor and move createException function

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29605 [rpc] refactor and move createException function**

Adds a wrapper around the existing `createException` function that
allows passing of an error string, instead of a regular C++ exception. This
allows us to mark futures with exceptions for errors that aren't necessarily c++
exceptions. This function is used by
https://github.com/pytorch/pytorch/pull/29601 and https://github.com/pytorch/pytorch/pull/26336.

Differential Revision: [D18439216](https://our.internmc.facebook.com/intern/diff/D18439216/)

rohan-varma committed:
[rpc] refactor and move createException function

Adds a wrapper around the existing createException function that
allows passing of an error string, instead of a regular C++ exception. This
allows us to createExceptions for errors that aren't necessarilu c++
exceptions. This function is used by
https://github.com/pytorch/pytorch/pull/29601 and
https://github.com/pytorch/pytorch/pull/26336.

Differential Revision: [D18439216](https://our.internmc.facebook.com/intern/diff/D18439216/)

[ghstack-poisoned]

zhaojuanmao reviewed (approved):
lgtm! please land it after tests are green

pritamdamania87 commented on torch/csrc/distributed/rpc/message.h:
```diff
@@ -115,6 +115,14 @@ class TORCH_API Message final {
 
+// create an exception given a message and exception.
+TORCH_API Message
+createException(const Message& request, const std::exception& e);
```
nit: Maybe we should call this `createExceptionResponse` and clearly state in the docs that we're building a `response` Message with an exception for the provided `request` Message.

rohan-varma committed:
Update on "[rpc] refactor and move createException function"


Adds a wrapper around the existing `createException` function that
allows passing of an error string, instead of a regular C++ exception. This
allows us to mark futures with exceptions for errors that aren't necessarily c++
exceptions. This function is used by
https://github.com/pytorch/pytorch/pull/29601 and https://github.com/pytorch/pytorch/pull/26336.

Differential Revision: [D18439216](https://our.internmc.facebook.com/intern/diff/D18439216/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] refactor and move createException function"


Adds a wrapper around the existing `createException` function that
allows passing of an error string, instead of a regular C++ exception. This
allows us to mark futures with exceptions for errors that aren't necessarily c++
exceptions. This function is used by
https://github.com/pytorch/pytorch/pull/29601 and https://github.com/pytorch/pytorch/pull/26336.

Differential Revision: [D18439216](https://our.internmc.facebook.com/intern/diff/D18439216/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] refactor and move createException function"


Adds a wrapper around the existing `createException` function that
allows passing of an error string, instead of a regular C++ exception. This
allows us to mark futures with exceptions for errors that aren't necessarily c++
exceptions. This function is used by
https://github.com/pytorch/pytorch/pull/29601 and https://github.com/pytorch/pytorch/pull/26336.

Differential Revision: [D18439216](https://our.internmc.facebook.com/intern/diff/D18439216/)

[ghstack-poisoned]

pietern commented on torch/csrc/distributed/rpc/message.cpp:
```diff
@@ -113,6 +113,24 @@ void Message::setId(int64_t id) {
+    const Message& request,
+    const std::exception& e) {
+  std::string exceptionMsg = e.what();
+  return createExceptionResponse(request, exceptionMsg);
```
Can you use `e.what()` inline?

rohan-varma committed:
Update on "[rpc] refactor and move createException function"


Adds a wrapper around the existing `createException` function that
allows passing of an error string, instead of a regular C++ exception. This
allows us to mark futures with exceptions for errors that aren't necessarily c++
exceptions. This function is used by
https://github.com/pytorch/pytorch/pull/29601 and https://github.com/pytorch/pytorch/pull/26336.

Differential Revision: [D18439216](https://our.internmc.facebook.com/intern/diff/D18439216/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/message.h:
```diff
@@ -115,6 +115,17 @@ class TORCH_API Message final {
+
+// Create a response Message with an exception type for the provided request
+// message. The passed in string will be used as the created message's payload
+TORCH_API Message createExceptionResponse(
```
It doesn't look like this method is used directly. Is there a reason we have this method? If not, can we remove it and only have one `createExceptionResponse` method for now?

rohan-varma replied:
I am planning on using them in the 2 PRs I have linked in the description, so I just added it here to remove it from the other 2.

rohan-varma committed:
Update on "[rpc] refactor and move createException function"


Adds a wrapper around the existing `createException` function that
allows passing of an error string, instead of a regular C++ exception. This
allows us to mark futures with exceptions for errors that aren't necessarily c++
exceptions. This function is used by
https://github.com/pytorch/pytorch/pull/29601 and https://github.com/pytorch/pytorch/pull/26336.

Differential Revision: [D18439216](https://our.internmc.facebook.com/intern/diff/D18439216/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] refactor and move createException function"


Adds a wrapper around the existing `createException` function that
allows passing of an error string, instead of a regular C++ exception. This
allows us to mark futures with exceptions for errors that aren't necessarily c++
exceptions. This function is used by
https://github.com/pytorch/pytorch/pull/29601 and https://github.com/pytorch/pytorch/pull/26336.

Differential Revision: [D18439216](https://our.internmc.facebook.com/intern/diff/D18439216/)

[ghstack-poisoned]

rohan-varma commented:
The CI failures are unrelated, will proceed with landing this
